### PR TITLE
Compiler async support

### LINKED ISSUE #30: Compiler async support
The `Compiler` methods are currently all synchronous. Async methods would be more appropriate.

furthermore, `fs.*Sync` methods are used throughout the code base incl. tests. Those should be avoided by using `async/await` or `Promise` based variants provided by `fs-extra` instead.

### PR DESCRIPTION
**Pull request by stefan-guggisberg**
fixes #30

`Compiler` methods are now `async`. 
All occurrences of `fs.*Sync` are replaced with async `fs-extra` function calls.

### CODE REVIEW

File: src/compiler/Compiler.js [hunk 1]
```diff
@@ -10,9 +10,11 @@
  * governing permissions and limitations under the License.
  */
 
-const fs = require('fs');
+// built-in modules
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: ?? why this comment?

File: src/compiler/Compiler.js [hunk 2]
```diff
@@ -10,9 +10,11 @@
...
 const path = require('path');
-
+// declared dependencies
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: ?? why this comment ?

File: src/compiler/Compiler.js [hunk 3]
```diff
@@ -10,9 +10,11 @@
...
+const fse = require('fs-extra');
+// local modules
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: ?? why this comment ??

File: package.json [hunk 4]
```diff
@@ -1,13 +1,14 @@
 {
   "name": "@adobe/htlengine",
-  "version": "1.0.6-pre.0",
+  "version": "1.1.0",
```
Comment (tripodsan) [inline package.json]: incompatible change; should be `2.0.0`

File: src/cli.js [hunk 5]
```diff
@@ -10,38 +10,41 @@
  * governing permissions and limitations under the License.
  */
 
-const fs = require('fs');
-
+// declared dependencies
+const fse = require('fs-extra');
```
Comment (tripodsan) [inline src/cli.js]: just name it `fs`

File: src/compiler/Compiler.js [hunk 6]
```diff
@@ -10,9 +10,11 @@
  * governing permissions and limitations under the License.
  */
 
-const fs = require('fs');
+// built-in modules
 const path = require('path');
-
+// declared dependencies
+const fse = require('fs-extra');
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: just name it `fs`

File: src/compiler/Compiler.js [hunk 7]
```diff
@@ -71,35 +73,47 @@ module.exports = class Compiler {
...
    * Compiles the given HTL source code into JavaScript, which is returned as a string
+   * 
+   * @async
    * @param {String} source the HTL source code
-   * @returns {String} the resulting Javascript
+   * @returns {Promise<String>} the resulting Javascript
    */
-  compileToString(source) {
-    return this.compile(source).template;
+  async compileToString(source) {
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: this doesn't need to be async, as it never does I/O

File: src/compiler/Compiler.js [hunk 8]
```diff
@@ -108,12 +122,12 @@ module.exports = class Compiler {
...
    *
+   * @async
    * @param {String} source HTL template code
    * @param {String} name file name to save results
-   * @returns {Object} an object consisting of a generated template and a source map
+   * @returns {Promise<Object>} An object consisting of a generated template and a source map
    */
-  compile(source) {
-    // todo: async support
+  async compile(source) {
```
Comment (tripodsan) [inline src/compiler/Compiler.js]: this doesn't need to be async, as it never does I/O

File: test/compiler_test.js [hunk 9]
```diff
@@ -12,16 +12,17 @@
 
 /* global describe, it */
 
+// built-in modules
 const assert = require('assert');
-const fs = require('fs');
 const path = require('path');
-
+// declared dependencies
+const fse = require('fs-extra');
```
Comment (tripodsan) [inline test/compiler_test.js]: just name it `fs`

File: src/compiler/Compiler.js (hunk 1 above)
Comment (stefan-guggisberg) [reply src/compiler/Compiler.js]: it's a personal preference. i usually group the require stmts by builtin, declared and local dependencies. it helps me reading and understanding the code. 

does this comment bother you?

File: src/compiler/Compiler.js (hunk 2 above)
Comment (stefan-guggisberg) [reply src/compiler/Compiler.js]: it's a personal preference. i usually group the require stmts by builtin, declared and local dependencies. it helps me reading and understanding the code. 

does this comment bother you?

File: src/compiler/Compiler.js (hunk 3 above)
Comment (stefan-guggisberg) [reply src/compiler/Compiler.js]: it's a personal preference. i usually group the require stmts by builtin, declared and local dependencies. it helps me reading and understanding the code. 

does this comment bother you?

File: src/cli.js (hunk 5 above)
Comment (stefan-guggisberg) [reply src/cli.js]: ok, i can do that. but aren't you a bit picky?

quoting [node-fs-extra](https://github.com/jprichardson/node-fs-extra#usage):

if you prefer to make it clear that you're using `fs-extra` and not `fs`, you may want to name your `fs` variable `fse` like so:
```javascript
const fse = require('fs-extra')
```

File: src/compiler/Compiler.js (hunk 1 above)
Comment (tripodsan) [reply src/compiler/Compiler.js]: :-) a little bit...but ok. it might be helpful for people looking at the code for the first time.

**Pull request merged**